### PR TITLE
Feat: Shops Rework

### PR DESCRIPTION
**Pull request by Zintixx**
- Implements handling of timed shops, rotating shops, and flexibility of account vs character-wide shops.
- Added proper item selling prices
- Currently there's still a lot of shops missing. You can test shops working with the following implemented shops:
  - Rotating equip shop: Bonin in Henesys
  - Timed shop: Rue Shop in Queenstown
  - Timed and account-wide shop: Bobby in Queenstown
 - Implements Buyback
   - There is a current known issue where if you try to npc a 13th item (12 is the max in buyback from official servers), instead of the list of items being pushed and the 12th item being unavailable to buyback, the 13th item takes the spot instead of being on the top left of the list.
   - #89

### CODE REVIEW

**Review by AlanMorel — APPROVED** [review]

**Review by Sparkymod — APPROVED** [review]

File: MapleServer2/Database/Classes/DatabaseShop.cs [hunk 1]
```diff
@@ -7,7 +8,7 @@ public class DatabaseShop : DatabaseTable
 {
     public DatabaseShop() : base("shops") { }
 
-    public Shop? FindById(int id)
+    public Shop FindById(int id)
```
Comment (AngeloTadeucci) [inline MapleServer2/Database/Classes/DatabaseShop.cs:11]: it has a return null inside so Shop can be null

File: MapleServer2/Database/Classes/DatabaseShop.cs [hunk 2]
```diff
@@ -19,8 +20,28 @@ public class DatabaseShop : DatabaseTable
...
+        return shops;
+    }
+
+    public void Update(Shop shop)
+    {
+        QueryFactory.Query(TableName).Where("id", shop.Id).Update(new
+        {
+            next_restock_timestamp = shop.RestockTime
+        });
+    }
```
Comment (AngeloTadeucci) [inline MapleServer2/Database/Classes/DatabaseShop.cs:41]: Either rename to (UpdateNextRestock or similar) or add all the parameters

File: MapleServer2/PacketHandlers/Game/SystemShopHandler.cs [hunk 3]
```diff
@@ -56,7 +58,9 @@ private static void HandleViaItem(GameSession session, PacketReader packet)
             return;
         }
 
-        ShopHelper.OpenSystemShop(session, item.Shop.ShopId, 0);
+        ShopHelper.OpenShop(session, item.Shop.ShopId, 0);
+        session.Send(SystemShopPacket.Open());
+
```
Comment (AngeloTadeucci) [inline MapleServer2/PacketHandlers/Game/SystemShopHandler.cs:63]:
```suggestion
```

File: MapleServer2/Types/BuyBackItem.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
+﻿using Maple2Storage.Enums;
+using MaplePacketLib2.Tools;
+using MapleServer2.Database;
+using MapleServer2.Packets.Helpers;
+using MySqlX.XDevAPI.CRUD;
```
Comment (AngeloTadeucci) [inline MapleServer2/Types/BuyBackItem.cs:5]: what this??

File: MapleServer2/Types/Player.cs [hunk 5]
```diff
@@ -152,6 +153,11 @@ public SubJobCode SubJobCode
     public Wallet Wallet { get; set; }
     public Dictionary<int, QuestStatus> QuestData;
 
+    public Dictionary<int, Shop> Shops = new();
+    public Dictionary<int, PlayerShopLog> ShopLogs = new();
+    public Dictionary<int, PlayerShopItemLog> ShopItemLogs = new();
```
Comment (AngeloTadeucci) [inline MapleServer2/Types/Player.cs:158]: Why we are storing logs in the player? 🤔

Comment (Zintixx) [reply MapleServer2/Types/Player.cs:158]: I figured this is easier than hitting the database everytime someone opens a shop up.

**Review by AngeloTadeucci — APPROVED** [review]

**Pull request merged**